[generator] Generator processors

maksimandrianov opened the pull request:
Коммит:
1. Added WritePOD, ReadPOD: Добавил функции, для записи простых типов. Далее они будут нужны в 2.
2. Added new serialization type: Добавил новый точный тип сериализации и десериализации, он нужен для записи во временные файлы, работает без потери точности.
3. Added affilations: Новый механизм принадлежности фичи регионам, далее в сл реквестах будет использован как processing layer.
4. Added processors: Появились процессоры, это сути старые эмиттеры. Посчитал, что название неудачное, потому что разных эмиттеров в коде достаточно и так. И все же теперь это больше похоже на processor, чем на эмиттер. Эмиттеры будут удалены позже.
5.  Old code support: Починил polygonizer, что бы код оставался рабочим. Поже polygonizer будет удален. Его роль будет выполнять affilations из коммита 3.

Звездочка: тут гитхаб сортирует коммиты по времени. Я делал ребейз что бы переупорядочить их. Так что git log даст правильный порядок

maksimandrianov commented:
JTALL

mpimenov commented on coding/read_write_utils.hpp:
```diff
@@ -25,6 +25,22 @@ namespace rw
   }
 
+  template <class TSink, class T>
+  void WritePOD(TSink & sink, T const & value)
```
https://github.com/mapsme/omim/blob/master/coding/write_to_sink.hpp

maksimandrianov replied:
template <class Sink, typename T>
std::enable_if_t<std::is_integral<T>::value || std::is_enum<T>::value, void> WriteToSink
а мне нужно для записи точек double

tatiana-yan replied:
кажется лучше WriteToSink дорабоать, а не делать другой метод который делает почти то же самое

mpimenov replied:
Да, тогда или WriteFloatingPointToSink, или вообще не тянуть это в coding/.
WriteToSink уже довольно опасна (всё было бы хорошо, если бы её проверка на endianness работала), не надо дальше ослаблять её требования.

maksimandrianov replied:
ок, перенс эти функции в анонимный нейспейс в feature_builder

mpimenov commented on coding/read_write_utils.hpp:
```diff
@@ -25,6 +25,22 @@ namespace rw
+  }
+
+  template <class TSource, class T>
+  void ReadPOD(TSource & src, T & value)
```
https://github.com/mapsme/omim/blob/master/coding/reader.hpp#L278

maksimandrianov replied:
++ удалил ReadPOD

mpimenov commented on generator/feature_builder.cpp:
```diff
@@ -444,6 +444,66 @@ void FeatureBuilder::DeserializeFromIntermediate(Buffer & data)
+    WriteVarInt(sink, m_coastCell);
+  }
+
+  // save OSM IDs to link meta information with sorted features later
```
s/save/Save/
s/later/later./

maksimandrianov replied:
++

mpimenov commented on generator/feature_builder.cpp:
```diff
@@ -444,6 +444,66 @@ void FeatureBuilder::DeserializeFromIntermediate(Buffer & data)
+
+  // save OSM IDs to link meta information with sorted features later
+  rw::WriteVectorOfPOD(sink, m_osmIds);
+  // check for correct serialization
```
s/check/Check/
s/serialization/serialization./

maksimandrianov replied:
++

mpimenov commented on generator/feature_builder.hpp:
```diff
@@ -231,51 +243,109 @@ class FeatureBuilder
+{
+enum class SerializationVersion : uint32_t
+{
+  Undef,
```
s/Undef/Undefined/

maksimandrianov replied:
++

mpimenov commented on generator/feature_builder.hpp:
```diff
@@ -231,51 +243,109 @@ class FeatureBuilder
+
 // Read feature from feature source.
-template <class Source>
+template <class SerializePolicy = serialization_policy::MinSize, class Source>
```
s/Serialize/Serialization/ ?

maksimandrianov replied:
++

mpimenov replied:
ping

maksimandrianov replied:
fixed

mpimenov commented on generator/feature_builder.hpp:
```diff
@@ -231,51 +243,109 @@ class FeatureBuilder
-
+//  TryReadAndCheckVersion<SerializePolicy>(src);
+  auto const fileSize = reader.Size();
+  uint64_t currPos = src.Pos();
```
auto?

maksimandrianov replied:
++

mpimenov commented on generator/feature_builder.hpp:
```diff
@@ -302,4 +372,35 @@ void ForEachParallelFromDatRawFormat(size_t threadsCount, std::string const & fi
+  FeatureBuilderWriter(std::string const & filename, FileWriter::Op op = FileWriter::Op::OP_WRITE_TRUNCATE)
+    : m_writer(filename, op)
+  {
+//    WriteVarUint(m_writer, static_cast<serialization_policy::TypeSerializationVersion>(SerializePolicy::kSerializationVersion));
```
add a comment why this is commented out, please

maksimandrianov replied:
++

mpimenov commented on generator/processor_coastline.hpp:
```diff
@@ -0,0 +1,34 @@
+
+namespace generator
+{
+// This class is implementation oft FeatureProcessorInterface for coastlines.
```
s/oft/of the/

mpimenov replied:
The comment looks superfluous.

maksimandrianov replied:
++

mpimenov commented on generator/processor_coastline.hpp:
```diff
@@ -0,0 +1,34 @@
+public:
+  explicit ProcessorCoastline(std::shared_ptr<FeatureProcessorQueue> const & queue);
+
+  // EmitterInterface overrides:
```
Are these EmitterInterface or FeatureProcessorInterface overrides?

maksimandrianov replied:
++

mpimenov commented on generator/processor_country.cpp:
```diff
@@ -0,0 +1,71 @@
+
+bool ProcessorCountry::Finish()
+{
+  //  WriteDump();
```
?

LaGrunge replied:
?

maksimandrianov replied:
это пригодится чуть позже, пока убрал

mpimenov commented on generator/processor_restaurants.hpp:
```diff
@@ -0,0 +1,48 @@
+    Stats operator+(Stats const & other) const
+    {
+      Stats s;
+      s.m_restaurantsPoi = m_restaurantsPoi  + other.m_restaurantsPoi;
```
clang-format

maksimandrianov replied:
++

mpimenov commented on generator/processor_simple.hpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace generator
+{
+// ProcessorSimpleWriter class is a simple emitter. It does not filter objects.
```
ProcessorSimple?

maksimandrianov replied:
++

mpimenov commented on generator/processor_simple.hpp:
```diff
@@ -0,0 +1,38 @@
+  explicit ProcessorSimple(std::shared_ptr<FeatureProcessorQueue> const & queue,
+                           std::string const & filename);
+
+  // EmitterInterface overrides:
```
s/EmitterInterface/FeatureProcessorInterface/ ?

maksimandrianov replied:
++

mpimenov commented on generator/processor_world.hpp:
```diff
@@ -0,0 +1,42 @@
+  explicit ProcessorWorld(std::shared_ptr<FeatureProcessorQueue> const & queue,
+                          std::string const & popularityFilename);
+
+  // EmitterInterface overrides:
```
FeatureProcessorInterface?

maksimandrianov replied:
++

mpimenov commented:
Чтобы оба порядка совпадали, можно сделать ```git commit --reset-author``` при ребейзе, как вариант.

tatiana-yan commented:
>Коммит:
>1. Added WritePOD, ReadPOD: Добавил функции, для записи простых типов. Далее они будут нужны в 2.
>2. Added new serialization type: Добавил новый точный тип сериализации и десериализации, он нужен для записи во временные файлы, работает без потери точности.
>3. Added affilations: Новый механизм принадлежности фичи регионам, далее в сл реквестах будет использован как processing layer.
>4. Added processors: Появились процессоры, это сути старые эмиттеры. Посчитал, что название неудачное, потому что разных эмиттеров в коде достаточно и так. И все же теперь это больше похоже на processor, чем на эмиттер. Эмиттеры будут удалены позже.
>5. Old code support: Починил polygonizer, что бы код оставался рабочим. Поже polygonizer будет удален. Его роль будет выполнять affilations из коммита 3.

тупой вопрос: нельзя ли всё это или часть из этого сделать отдельными реквестами?

maksimandrianov commented:
> > Коммит:
> > 
> > 1. Added WritePOD, ReadPOD: Добавил функции, для записи простых типов. Далее они будут нужны в 2.
> > 2. Added new serialization type: Добавил новый точный тип сериализации и десериализации, он нужен для записи во временные файлы, работает без потери точности.
> > 3. Added affilations: Новый механизм принадлежности фичи регионам, далее в сл реквестах будет использован как processing layer.
> > 4. Added processors: Появились процессоры, это сути старые эмиттеры. Посчитал, что название неудачное, потому что разных эмиттеров в коде достаточно и так. И все же теперь это больше похоже на processor, чем на эмиттер. Эмиттеры будут удалены позже.
> > 5. Old code support: Починил polygonizer, что бы код оставался рабочим. Поже polygonizer будет удален. Его роль будет выполнять affilations из коммита 3.
> 
> тупой вопрос: нельзя ли всё это или часть из этого сделать отдельными реквестами?

можно но нужно ли? я хотел, бы успеть все сделать до релиза, и так все разбираю реквесты, вся работа примерно 7к изменных строк была. 
и кажется что это логичный реквест, добавились процессоры, точнее переместились(раньше это были эмиттеры), а все остальные изменения связаны с их добавлением

tatiana-yan commented:
ревью реквеста на 1000 строк занимает больше времени, чем ревью 5 реквестов по 200 строк

tatiana-yan commented on generator/processor_world.hpp:
```diff
@@ -0,0 +1,42 @@
+
+namespace generator
+{
+// This class is implementation of EmitterInterface for the world.
```
FeatureProcessorInterface?

maksimandrianov replied:
++

tatiana-yan commented on generator/feature_builder.hpp:
```diff
@@ -134,6 +135,12 @@ class FeatureBuilder
   }
 
+  template <class FnT>
+  bool HasTypesIf(FnT fn) const
```
AnyOfTypes?

в этом реквесте не используется

maksimandrianov replied:
++ пока убрал

tatiana-yan commented on generator/feature_builder.hpp:
```diff
@@ -302,4 +372,35 @@ void ForEachParallelFromDatRawFormat(size_t threadsCount, std::string const & fi
+class FeatureBuilderWriter
+{
+public:
+  FeatureBuilderWriter(std::string const & filename, FileWriter::Op op = FileWriter::Op::OP_WRITE_TRUNCATE)
```
explicit?

maksimandrianov replied:
++

tatiana-yan commented on generator/feature_builder.hpp:
```diff
@@ -178,6 +185,10 @@ class FeatureBuilder
 
+  // These methods use geometry without loss of accuracy.
+  void SerializeAccuratelyForIntermediate(Buffer & data) const;
+  void DeserializeAccuratelyFromIntermediate(Buffer & data);
```
тесты что после сериализации/десериализации получаем исходный результат?

maksimandrianov replied:
++

tatiana-yan commented:
тесты coding на запись/чтение double?

tatiana-yan commented on generator/borders.hpp:
```diff
@@ -86,4 +139,14 @@ std::vector<m2::RegionD> ReadPolygonsOfOneBorder(Source & src)
+class PackedBorders
+{
+public:
+  static CountriesContainer const & GetOrCreate(std::string const & name);
```
GetOrLoad?

maksimandrianov replied:
все же объект создается, давай оставим GetOrCreate

tatiana-yan replied:
Ты передаешь в метод путь откуда нужно читать данные, метод по этому пути что-то читает и заполняет структуры класса этими данными.
Для меня это скорее Load, но если тебе так понятнее Create -- ок.

tatiana-yan commented on generator/borders.hpp:
```diff
@@ -86,4 +139,14 @@ std::vector<m2::RegionD> ReadPolygonsOfOneBorder(Source & src)
+
+private:
+  static std::mutex m_mutex;
+  static std::unordered_map<std::string, CountriesContainer> m_countries;
```
в норме в этом мэпе может быть больше одного элемента? я правильно понимаю что ключ -- это путь к директории с файлами .poly ?

maksimandrianov replied:
да, сейчас эта возможность не используется, но можно представить что в бордерсах появятся дирректории - страны

tatiana-yan replied:
понятно, спасибо за пояснения

tatiana-yan commented on generator/borders.cpp:
```diff
@@ -287,4 +294,22 @@ void UnpackBorders(std::string const & baseDir, std::string const & targetDir)
+  CountriesContainer countries;
+  CHECK(LoadCountriesList(name, countries), ("Error loading country polygons files."));
+  m_countries.emplace(name, countries);
+  return m_countries[name];
```
emplace вернёт пару, в которой уже есть итератор, указывающий на вставленный элемент.

тут конечно map небольшой, но в общем случае это лишний поиск

maksimandrianov replied:
в общем да, но это ще хэш таблица
плюс, кажется, что он вернут пару, где будет копия, а не ссылка

tatiana-yan replied:
в паре итератор, разыменование итератора дает ссылку

maksimandrianov replied:
++

tatiana-yan commented on generator/borders.cpp:
```diff
@@ -287,4 +294,22 @@ void UnpackBorders(std::string const & baseDir, std::string const & targetDir)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  if (m_countries.count(name) != 0)
+    return m_countries[name];
```
тут тоже можно одним поиском в map обойтись, но up to you

maksimandrianov replied:
кажется, что он вернет пару, где будет копия, а не ссылка

tatiana-yan replied:
```C++
auto const it = m_countries.find(name);
if (it != m_countrie.end())
  return *it;
```
какая копия?

maksimandrianov replied:
++ it в map будет парой и ключа и значения

tatiana-yan commented on generator/borders.hpp:
```diff
@@ -39,24 +42,74 @@ namespace borders
+
+private:
+  m4::Tree<CountryPolygons> m_regionsTree;
+  std::unordered_map<std::string, CountryPolygons> m_regions;
```
я правильно понимаю что CountryPolygons довольно тяжелая структура и теперь мы храним все CountryPolygons по два раза?

нельзя ли в качестве значения в m_regionsTree хранить более лёгкое имя и находить CountryPolygons по имени?

maksimandrianov replied:
они будут храниться только здесь, а в другим местах будут ссылки
эта структура как раз сделана для того, что бы хранить один экземпляр на запуск генератора

tatiana-yan replied:
Но ведь мы прям здесь в m_regionsTree и m_regions мы будем хранить CountryPolygons по два раза.

maksimandrianov replied:
ой, я думал ты про другое место говорила вообще, сорри, могу сделать ссылки где нибудь, например в m_regionsTree

maksimandrianov replied:
поправил

LaGrunge commented on generator/affiliation.cpp:
```diff
@@ -0,0 +1,62 @@
+  {
+    for (borders::CountryPolygons const & countryPolygons : countriesContainer)
+    {
+      auto const need = fb.ForAnyGeometryPoint([&](auto const & point) {
```
OMG это же bool

maksimandrianov replied:
не делал в одну строку что бы читалось лучше

LaGrunge commented on generator/processor_factory.hpp:
```diff
@@ -0,0 +1,51 @@
+{
+  switch (type)
+  {
+  case ProcessorType::Coastline:
```
nit: Можно фигануть макросом. Можно и не.

maksimandrianov replied:
макрос не хочется

LaGrunge commented on generator/processor_world.cpp:
```diff
@@ -0,0 +1,54 @@
+  m_affilationsLayer->AddBufferToQueue(m_queue);
+}
+
+bool ProcessorWorld::Finish()
```
А какой-то нетривиальный Finish будет у процессоров? И что если он вернет false?

maksimandrianov replied:
++ сейчас уже нет, переносил код, осталось, убрал

tatiana-yan commented on generator/borders.hpp:
```diff
@@ -39,24 +42,74 @@ namespace borders
+  }
+
+  CountryPolygons const & GetRegionByName(std::string const & name) const
+  {
```
ASSERT HasRegionByName ?

maksimandrianov replied:
++ добавил, но ничего страшного не было бы, там at() который бросит исключение)

tatiana-yan commented on generator/affiliation.hpp:
```diff
@@ -0,0 +1,48 @@
+public:
+  virtual ~AffiliationInterface() = default;
+
+  virtual std::vector<std::string> GetAffiliations(FeatureBuilder const & fb) const = 0;
```
добавь пожалуйста описание к методу

maksimandrianov replied:
++

tatiana-yan commented on generator/affiliation.hpp:
```diff
@@ -0,0 +1,48 @@
+public:
+  virtual ~AffiliationInterface() = default;
+
+  virtual std::vector<std::string> GetAffiliations(FeatureBuilder const & fb) const = 0;
```
может тут не std::string, в storage::CountryId использовать? как дальше будет использоваться результат GetAffiliations?

maksimandrianov replied:
не знаю, а что такое storage::CountryId?

maksimandrianov replied:
это имя границы, куда будет записан FeatureBuilder

tatiana-yan replied:
storage::CountryId это тип для идентификаторов стран (которые совпадают с названиями mwm и ключами для листьев в CountryTree/countries.txt)
это string, но при чтении сразу понятно что тут речь идёт о названиях mwm и всём таком

maksimandrianov replied:
аа, в моем слкчае это более широкое понятие
пусть лучше будет асбтрактная строка, потому что для b2b это тоже будет использоваться, но там это не country id, что может запутать и кстати сейчас в polygonizer это тоже не CountryId
я написал в комменте, что функция вернет имена бакетов, которым принадлежит фича:)

tatiana-yan replied:
>я написал в комменте, что функция вернет имена бакетов, которым принадлежит фича:)

но это неправда, для OneFileAffilation она вернёт имя этого файла, даже если фича ему не принадлежит

maksimandrianov replied:
технически же она ему принадлежит
даже название OneFileAffilation
если кажется что запутанно, предложи, как лучше написать поменяю

tatiana-yan commented on generator/affiliation.hpp:
```diff
@@ -0,0 +1,48 @@
+
+private:
+  borders::CountriesContainer const & m_countries;
+  bool m_isMwmsForWholeWorld;
```
m_haveBordersForWholeWorld ?

maksimandrianov replied:
++

tatiana-yan commented on generator/affiliation.cpp:
```diff
@@ -0,0 +1,62 @@
+  {
+    borders::CountryPolygons const & countryPolygons= countriesContainer.front();
+    countries.emplace_back(countryPolygons.GetName());
+    return countries;
```
Например у нас фича находится в воде и не принадлежит никакой стране (например, это море).
Но тут она может попасть в грубый контур какой-то страны, а дальше мы не проверим, попадает ли в точный.
Это ок?

maksimandrianov replied:
эта оптимизация, которая была раньше, на само деле она очень классная
проверил, море в mwm ки не складывается, только в world

tatiana-yan replied:
а как именно ты это проверил?
а не моря?
можно поставить тут ASSERT что фича попадает в контур и прогнать генерацию на маленьком мире в дебаге например?

maksimandrianov replied:
то же самое крутилось в релизе года 2 https://github.com/mapsme/omim/blob/release-91/generator/polygonizer.hpp#L116
генерировал старым генератором и новым и сравнивал все фичи, которые попадают
ок проверю

tatiana-yan replied:
То что всегда так делали не значит что это правильно.
Например у нас есть проблема что у некоторых mwm необъяснимо большие limit rect, может её причина в этом (а может нет).
Я понимаю экономию в вычислениях, но хочется понимать и вносит ли эта экономия ошибки и, если да, то какие. А дальше уже решать ок нам эти ошибки или не ок.

maksimandrianov replied:
можно я не буду этот вопрос исследовать в рамках этой работы? ее цель не в поиске существующих ошибок а в организации кода так, что бы это можно было распараллелить, плюс это не новая функциональность
если так цепляться то можно и весь проект переписать:)

tatiana-yan replied:
а где этот код сейчас? (можно ссылку?)

если сейчас так работает -- ок, давай исследование отдельно, но в рамках этого реквеста оставь пожалуйста тут todo

tatiana-yan replied:
в polygonizer проверяется нахождение точки внутри границ страны независимо от кол-ва найденных стран:

```C++
   void operator()(FeatureBuilder & fb)
   {
     m_countries.ForEachInRect(fb.GetLimitRect(), [&](auto const & countryPolygons) {
       auto const need = fb.ForAnyGeometryPoint([&](auto const & point) {
         auto const & regions = countryPolygons.m_regions;
         return regions.ForAnyInRect(m2::RectD(point, point), [&](auto const & rgn) {
           return rgn.Contains(point);
         });
       });
 
       if (need)
         this->EmitFeature(countryPolygons, fb);
     });
   }
```

tatiana-yan replied:
ping

maksimandrianov replied:
++ написал todo, что оптимизация требует исследования, что есть гипотеза, про которую ты сказала
сейчас этого нет в polygonizer в мастере, потому что я это удалял эту оптимизацию вместе с параллельностью которая там была, но есть в релизе(ссылку скинул выше). сейчас она будет опциональной и применяться с указанием флага, что собираем карты для всего мира. убрал я ее из-за проблемы пересборки части mwmок, после того как собрали карты для всего мира.
пример: букинг не добавился в италию, хотим пересобрать только ее. раньше не могли потому что в таком случае границы будут неточные, после пересборки получатся не наши mwmки, а какая то фигня.

tatiana-yan commented on generator/processor_world.hpp:
```diff
@@ -0,0 +1,42 @@
+  void MergeInto(ProcessorWorld & other) const override;
+
+private:
+  std::string m_popularityFilename;
```
как используется это поле?

tatiana-yan commented on generator/processor_factory.hpp:
```diff
@@ -0,0 +1,51 @@
+#include "generator/processor_restaurants.hpp"
+#include "generator/processor_simple.hpp"
+#include "generator/processor_world.hpp"
+#include "generator/factory_utils.hpp"
```
отсортируй пожалуйста

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_factory.hpp:
```diff
@@ -0,0 +1,51 @@
+  Coastline,
+  World,
+  Noop
+  //  Booking
```
а почему букинг закомментирован?

maksimandrianov replied:
он и был закоментиован в эмитерах, потому что он не испольуется + я его еще тогда не перевел, когда были эмиттеры, были некоторые сложности с шаблонным кодом который используется в обработке Booking

tatiana-yan commented on generator/processor_noop.hpp:
```diff
@@ -0,0 +1,31 @@
+
+namespace generator
+{
+class ProcessorNoop : public FeatureProcessorInterface
```
где будет использоваться?

maksimandrianov replied:
в тестах

maksimandrianov replied:
посмотри плиз, сейчас есть EmitterNoop
можно сказть это просто перименование с некоторыми небольшими доработками внутри классов

tatiana-yan replied:
если только в тестах и нигде больше не планируется, можно в директорию тестов перенести

maksimandrianov replied:
на базе него можно данные собирать и статистику всякую
лучше не надо

maksimandrianov replied:
плюс он создается как и все процессоры через фабрику

tatiana-yan replied:
ок

tatiana-yan commented on generator/processor_noop.hpp:
```diff
@@ -0,0 +1,31 @@
+
+#include <memory>
+#include <string>
+#include <vector>
```
string и vector не нужны

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_booking.hpp:
```diff
@@ -0,0 +1,56 @@
+class ProcessorBooking : public FeatureProcessorInterface
+{
+public:
+  ProcessorBooking(Dataset const & dataset, map<base::GeoObjectId, feature::FeatureBuilder> & features)
```
s/map/std::map

maksimandrianov replied:
++

tatiana-yan commented:
в реквесте добавляется 1000 строк нового кода, куча новых классов и нет тестов на них
конечно unit test-ы не избавляют от всех ошибок, но от части избавляют
не хотелось бы это тестировать в процессе сборки новых карт к релизу

tatiana-yan commented on generator/processor_booking.hpp:
```diff
@@ -0,0 +1,56 @@
+    : m_dataset(dataset), m_features(features) {}
+
+  // FeatureProcessorInterface overrides:
+  virtual std::shared_ptr<FeatureProcessorInterface> Clone() const
```
include memory

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_booking.hpp:
```diff
@@ -0,0 +1,56 @@
+  // FeatureProcessorInterface overrides:
+  virtual std::shared_ptr<FeatureProcessorInterface> Clone() const
+  {
+    CHECK(false, ());
```
нет ворнинга компилятора из-за отсутствия return?

maksimandrianov replied:
++ сделал return {}

tatiana-yan commented on generator/processor_booking.hpp:
```diff
@@ -0,0 +1,56 @@
+
+private:
+  Dataset const & m_dataset;
+  map<base::GeoObjectId, feature::FeatureBuilder> & m_features;
```
std::

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_coastline.hpp:
```diff
@@ -0,0 +1,33 @@
+
+#include <memory>
+#include <string>
+#include <vector>
```
string и vector не используются

maksimandrianov replied:
++ убрал

tatiana-yan commented on generator/processor_country.hpp:
```diff
@@ -0,0 +1,45 @@
+  std::shared_ptr<AffilationsFeatureLayer<>> m_affilationsLayer;
+  std::shared_ptr<FeatureProcessorQueue> m_queue;
+  std::shared_ptr<LayerBase> m_processingChain;
+  bool m_isMwmsForWholeWorld;
```
m_haveBordersForWholeWorld?

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_restaurants.cpp:
```diff
@@ -0,0 +1,59 @@
+  {
+  case feature::GeomType::Point: ++m_stats.m_restaurantsPoi; break;
+  case feature::GeomType::Area: ++m_stats.m_restaurantsBuilding; break;
+  default: ++m_stats.m_unexpectedFeatures;
```
сейчас у нас часть unexpectedFeatures получается обработана, а часть нет, это странно

maksimandrianov replied:
++ добавил ветку else

tatiana-yan commented on generator/processor_restaurants.hpp:
```diff
@@ -0,0 +1,47 @@
+#include "generator/processor_interface.hpp"
```
pragma once ?

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_restaurants.hpp:
```diff
@@ -0,0 +1,47 @@
+#include "generator/processor_interface.hpp"
+#include "generator/feature_builder.hpp"
```
отсортируй пожалуйста

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_restaurants.hpp:
```diff
@@ -0,0 +1,47 @@
+  explicit ProcessorRestaurants(std::shared_ptr<FeatureProcessorQueue> const & queue);
+
+  // FeatureProcessorInterface overrides:
+  std::shared_ptr<FeatureProcessorInterface> Clone() const override;
```
include memory

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_restaurants.hpp:
```diff
@@ -0,0 +1,47 @@
+    // Number of features of any "food type".
+    uint32_t m_restaurantsPoi = 0;
+    uint32_t m_restaurantsBuilding = 0;
+    uint32_t m_unexpectedFeatures = 0;
```
include cstdint

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_simple.hpp:
```diff
@@ -0,0 +1,38 @@
+
+#include "generator/feature_builder.hpp"
+#include "generator/feature_processing_layers.hpp"
+#include "generator/processor_interface.hpp"
```
отсортируй пожалуйста

maksimandrianov replied:
тут вроде ок

tatiana-yan replied:
да, извини, тут ок)

tatiana-yan commented on generator/processor_simple.hpp:
```diff
@@ -0,0 +1,38 @@
+{
+public:
+  explicit ProcessorSimple(std::shared_ptr<FeatureProcessorQueue> const & queue,
+                           std::string const & filename);
```
предлагаю filename назвать как-то более говоряще, чтобы было ясно что это за файл

maksimandrianov replied:
предложи как, пожалуйста

tatiana-yan replied:
так как я не поняла что делает OneFileAffilation (обсуждаем это с тобой выше), я не могу понять что это за файл:)

tatiana-yan replied:
mwmName?

maksimandrianov replied:
после изменения имени класса это еще имеет акутальность?

maksimandrianov replied:
потрело акутальность после переименования Affilation?

tatiana-yan replied:
глядя на конструктор по-прежнему не понятно какое имя сюда нужно передать
и что это за имя? полный путь к какому-то файлу? имя файла мвм? имя файла мвм с расширением? имя файла mwm без расширения?

tatiana-yan replied:
ping

maksimandrianov replied:
++ сделал просто name, написал комментарий для того, что бы было понятно

tatiana-yan commented on generator/affiliation.hpp:
```diff
@@ -0,0 +1,49 @@
+  bool m_haveBordersForWholeWorld;
+};
+
+class OneFileAffiliation : public AffiliationInterface
```
WholeWorldAffiliation?

maksimandrianov replied:
думаешь так понятнее? ну ок

maksimandrianov replied:
просто ассоциации со словом World немного другие

tatiana-yan replied:
можно ещё подумать, onefile тоже не понятно

tatiana-yan replied:
SingleAffiliation ?

maksimandrianov replied:
++ SingleAffiliation круто, так лучше

mpimenov commented on generator/affiliation.cpp:
```diff
@@ -0,0 +1,58 @@
+
+namespace feature
+{
+CountriesFilesAffiliation::CountriesFilesAffiliation(std::string const & borderPath, bool haveBordersForWholeWorld)
```
Насколько здесь важно именно Affiliation вместо, скажем, Membership?
Мы уже используем в countries.txt термин Affiliation в другом смысле.

maksimandrianov replied:
в там разве используется не в смысле принадлежности?

mpimenov replied:
Там это принадлежность и синонимы для регионов, насколько я помню.
Главное отличие в типах: там аффилированы регион <-> регион, а здесь регион <-> фича.

maksimandrianov replied:
в генераторе по affiliation ничего не находится кроме этих классов, считаю что они никак не перекрываются, что бы их менять
предлагаю оставить

mpimenov commented on generator/borders.hpp:
```diff
@@ -39,24 +42,89 @@ namespace borders
+
+    CountryPolygons const * country;
+    m_regionsTree.FindNode([&](auto const & countryPolygons) {
+      country = &countryPolygons;
```
мб присваивать только при совпадении?
заодно можно будет избавиться от лишнего вызова в ASSERT

maksimandrianov replied:
++ ASSER все же оставлю, иначе придется проверять что вообще было что-то найдено уже после выполнения каких-то действий, на мой взгляд лучше проверить до

mpimenov commented on generator/feature_builder.cpp:
```diff
@@ -51,6 +53,22 @@ bool IsEqual(vector<m2::PointD> const & v1, vector<m2::PointD> const & v2)
                [](m2::PointD const & p1, m2::PointD const & p2) { return IsEqual(p1, p2); });
 }
+
+template <class TSink, class T>
```
s/TSink/Sink/g

maksimandrianov replied:
++

mpimenov commented on generator/feature_builder.cpp:
```diff
@@ -444,6 +462,66 @@ void FeatureBuilder::DeserializeFromIntermediate(Buffer & data)
+  {
+    m_polygons.clear();
+    uint32_t const count = ReadVarUint<uint32_t>(source);
+    ASSERT_GREATER (count, 0, (*this));
```
clang-format

maksimandrianov replied:
++

mpimenov commented on generator/processor_factory.hpp:
```diff
@@ -0,0 +1,50 @@
+#include "generator/processor_restaurants.hpp"
+#include "generator/processor_simple.hpp"
+#include "generator/processor_world.hpp"
+#include "base/assert.hpp"
```
пустая строка

maksimandrianov replied:
++

mpimenov commented on generator/processor_restaurants.hpp:
```diff
@@ -0,0 +1,49 @@
+    {
+      Stats s;
+      s.m_restaurantsPoi = m_restaurantsPoi + other.m_restaurantsPoi;
+      s.m_restaurantsBuilding = m_restaurantsBuilding  + other.m_restaurantsBuilding;
```
clang-format

maksimandrianov replied:
++

tatiana-yan commented on generator/affiliation.hpp:
```diff
@@ -0,0 +1,47 @@
+  CountriesFilesAffiliation(std::string const & borderPath, bool haveBordersForWholeWorld);
+
+  // AffiliationInterface overrides:
+  // The method will return the names of the buckets to which the fb belongs.
```
может в базовый класс описание?

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_restaurants.cpp:
```diff
@@ -0,0 +1,59 @@
+  : m_queue(queue)
+{
+  auto affilation = std::make_shared<feature::OneFileAffiliation>("Restaurants");
+  m_affilationsLayer = std::make_shared<AffilationsFeatureLayer<>>(kAffilationsBufferSize, affilation);
```
почему у ProcessorRestaurants есть m_affilationsLayer, а у ProcessorBooking и ProcessorNoop нет? от чего зависит, нужен ли m_affilationsLayer?

maksimandrianov replied:
m_affilationsLayer - узнает принадлежность и записывает элемент в очередь, т.е если нужны обработанне элементы они могут быть взяты из очереди.
для ProcessorBooking и ProcessorNoop не нужно смотреть на обработанные элементы

tatiana-yan replied:
можно какую-то блок-схему? устно/на доске/в комментах/как угодно

```если нужны обработанне элементы они могут быть взяты из очереди```
кем обработаны элементы? кто их берет из очереди?
а то не ясно в чем тут принципиальное отличие ресторанов от отелей и почему одни надо класть в очередь, а другие нет

maksimandrianov replied:
думаюнужно выпилить ProcessorRestaurants и все что с этим связано в сл реквесте, т.к вопросов вызывает много, но не ипользуется, осталось от предков и волочится за собой до сих пор

tatiana-yan replied:
оставь пожалуйста в этом реквесте todo, либо, если в следующем реквесте его планируется выпилить, можно просто в этом его не добавлять :)

maksimandrianov replied:
да в этом выпилю пожалуй processor

tatiana-yan commented on generator/processor_restaurants.cpp:
```diff
@@ -0,0 +1,59 @@
+    {
+    case feature::GeomType::Point: ++m_stats.m_restaurantsPoi; break;
+    case feature::GeomType::Area: ++m_stats.m_restaurantsBuilding; break;
+    default: ++m_stats.m_unexpectedFeatures;
```
в итоге если у нас фича записана в  unexpected features из-за неправильных типов или отсутствия имени она обработана дальше не будет (не будет передана в m_processingChain и т.п.)

а если фича записана в unexpected features из-за неправильного геометрического типа, то она будет дальше обработана (передана в m_processingChain и т.п.)

так и должно быть?

maksimandrianov replied:
https://github.com/mapsme/omim/blob/release-80/generator/restaurants_info/restaurants_info.cpp#L52
верну назад, раньше было правильно

tatiana-yan replied:
давай устно обсудим, а то я не понимаю что является критерием правильности

maksimandrianov replied:
то что задумали авторы этого кода, я оставил логику:)

tatiana-yan replied:
ок

tatiana-yan commented on generator/processor_simple.hpp:
```diff
@@ -0,0 +1,38 @@
+
+#include <memory>
+#include <string>
+#include <vector>
```
вектор не нужен

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_world.hpp:
```diff
@@ -0,0 +1,41 @@
+
+private:
+  std::string m_popularityFilename;
+  std::shared_ptr<AffilationsFeatureLayer<>> m_affilationsLayer;
```
что такое AffilationsFeatureLayer ? он появится в следующем реквесте?

maksimandrianov replied:
скорее всего да, что это такое написал выше

tatiana-yan commented on generator/processor_booking.hpp:
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include "generator/processor_interface.hpp"
+#include "generator/feature_builder.hpp"
```
отсортируй пожалуйста

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_booking.hpp:
```diff
@@ -0,0 +1,51 @@
+
+#include "base/assert.hpp"
+#include "base/geo_object_id.hpp"
+#include "base/logging.hpp"
```
вроде не используется

maksimandrianov replied:
++ убрал

tatiana-yan commented on generator/generator_tests/feature_builder_test.cpp:
```diff
@@ -63,6 +63,12 @@ UNIT_CLASS_TEST(TestWithClassificator, FBuilder_ManyTypes)
   TEST_EQUAL(fb1, fb2, ());
   TEST_EQUAL(fb2.GetTypesCount(), 6, ());
+
+  fb1.SerializeAccuratelyForIntermediate(buffer);
```
эта сериализация от обычной отличается тем, что она сериализует геометрию без потерь.

в этом сравнении для сравнения геометрии используется 
```
base::AlmostEqualAbs(d1, d2, kMwmPointAccuracy);
```
то есть, то что геометрия сериализовывается без потерь не проверяется никак

предлагаю добавить отдельный тест на сериализацию-десериалицазию   FeatureBuilder с проверкой геометрии без kMwmPointAccuracy

maksimandrianov replied:
++ да спасибо, я не обратил на это внимание, добавил метод который проверяет на точное развенство и выделил в отдельный тест

tatiana-yan replied:
спасибо!

tatiana-yan commented on generator/processor_country.cpp:
```diff
@@ -0,0 +1,65 @@
+
+#include <fstream>
+
+#include "defines.hpp"
```
зачем тут?

maksimandrianov replied:
убрал, осталось от старого кода

tatiana-yan commented on generator/processor_world.hpp:
```diff
@@ -0,0 +1,41 @@
+
+#include <memory>
+#include <string>
+#include <vector>
```
не нужен

maksimandrianov replied:
++ убрал

tatiana-yan commented on generator/feature_builder.hpp:
```diff
@@ -43,7 +44,8 @@ class FeatureBuilder
   FeatureBuilder();
-  bool operator==(FeatureBuilder const &) const;
+  bool operator==(FeatureBuilder const & fb) const;
+  bool IsExactEq(FeatureBuilder const & fb) const;
```
добавь пожалуйста комментарий об отличии этого метода и оператора ==

maksimandrianov replied:
Exact трубет комментария?

tatiana-yan replied:
давай выберем произвольного разработчика из команды, спросим чем, судя по имени, отличаются эти методы и, если он угадает, то коммент не нужен:)

maksimandrianov replied:
++ написал

tatiana-yan replied:
спасибо!

tatiana-yan commented on generator/generator_tests/feature_builder_test.cpp:
```diff
@@ -271,3 +271,38 @@ UNIT_CLASS_TEST(TestWithClassificator, FeatureBuilder12_SerializeLocalityObjectF
+  params.FinishAddingTypes();
+  fb1.SetParams(params);
+
+  fb1.AddPoint(m2::PointD(0, 0));
```
мне кажется тут нужна геометрия посложнее, (0,0) сериализовать без потерь не сложно

maksimandrianov replied:
++ сделал посложнее

tatiana-yan replied:
спасибо!

tatiana-yan commented:
https://github.com/mapsme/omim/pull/11153#discussion_r300659601

ping

maksimandrianov commented:
@mpimenov , @tatiana-yan PTAL

tatiana-yan commented on generator/translator.cpp:
```diff
@@ -30,6 +30,9 @@ Translator::Translator(std::shared_ptr<EmitterInterface> emitter, cache::Interme
 
 void Translator::Emit(OsmElement & element)
 {
+  if (element.m_id != 8454075)
```
?

maksimandrianov replied:
++ убрал, спасибо

mpimenov commented on generator/borders.hpp:
```diff
@@ -11,7 +11,10 @@
 
 #include <string>
```
?

maksimandrianov replied:
++

mpimenov commented on generator/borders.hpp:
```diff
@@ -39,24 +42,96 @@ namespace borders
+        country = &countryPolygons;
+        return true;
+      }
+      else
```
no else after return

maksimandrianov replied:
++

mpimenov commented on generator/processor_simple.hpp:
```diff
@@ -0,0 +1,37 @@
+class ProcessorSimple : public FeatureProcessorInterface
+{
+public:
+  // name is backet name. For example it may "World", "geo_objects", "regions" etc.
```
s/name/|name|/
s/backet/bucket/ ?

mpimenov replied:
s/may/may be/

maksimandrianov replied:
++

mpimenov commented on generator/processor_world.cpp:
```diff
@@ -0,0 +1,49 @@
+
+std::shared_ptr<FeatureProcessorInterface> ProcessorWorld::Clone() const
+{
+   return std::make_shared<ProcessorWorld>(m_queue, m_popularityFilename);
```
clang-format

maksimandrianov replied:
++

tatiana-yan commented on generator/processor_simple.hpp:
```diff
@@ -6,31 +6,30 @@
 class ProcessorSimple : public FeatureProcessorInterface
 {
 public:
+  // |name| is bucket name. For example it may "World", "geo_objects", "regions" etc.
```
may be

maksimandrianov replied:
++
